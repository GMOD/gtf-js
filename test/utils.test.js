import gtf from '../src'

const {
  parseAttributes,
  parseFeature,
  parseDirective,
  formatFeature,
  escapeColumn,
} = gtf.util

describe('GTF utils', () => {
  it('can escape/unescape properly', () => {
    expect(gtf.util.escape(5)).toEqual('5')
    expect(gtf.util.unescape(' ')).toEqual(' ')
    expect(gtf.util.unescape(5)).toEqual('5')
    expect(escapeColumn('Noggin,+-%Foo\tbar')).toEqual('Noggin,+-%25Foo%09bar')
  })
  it('test parsing the attributes of a gtf line', () => {
    const testAttributes = parseAttributes(
      'transcript_id "EDEN.3"; gene_id "EDEN"; gene_name "EDEN";',
    )
    // eslint-disable-next-line camelcase
    const { gene_id, transcript_id } = testAttributes
    expect(gene_id[0]).toBe('EDEN')
    expect(transcript_id[0]).toBe('EDEN.3')
  })
  it('test parsing/formating a gtf feature', () => {
    const featureLine =
      'ctgA	example	CDS	3301	3902	.	+	0	transcript_id "EDEN.3"; gene_id "EDEN"; gene_name "EDEN";'
    const testFeature = parseFeature(featureLine)
    expect(testFeature).toMatchSnapshot()
    const featureToFormat = {
      seq_name: 'ctgA',
      source: 'example',
      feature: 'CDS',
      start: 3301,
      end: 3902,
      score: null,
      strand: '+',
      frame: 0,
      attributes: {
        transcript_id: 'EDEN.3',
        gene_id: 'EDEN',
        gene_name: 'EDEN',
      },
    }
    const formattedFeature = formatFeature(featureToFormat)
    expect(formattedFeature.trim()).toBe(featureLine.replace(/["]+/g, ''))
  })
  it('test parsing a gtf directive', () => {
    const testDirective = parseDirective('#!genome-build GRCh38.p7')
    expect(testDirective.directive).toBe('!genome-build')
  })
})
