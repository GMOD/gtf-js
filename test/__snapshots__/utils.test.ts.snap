// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GTF utils test parsing/formating a gtf feature 1`] = `
Object {
  "attributes": Object {
    "gene_id": Array [
      "\\"EDEN\\"",
    ],
    "gene_name": Array [
      "\\"EDEN\\"",
    ],
    "transcript_id": Array [
      "\\"EDEN.3\\"",
    ],
  },
  "end": 3902,
  "featureType": "CDS",
  "frame": "0",
  "score": null,
  "seq_name": "ctgA",
  "source": "example",
  "start": 3301,
  "strand": "+",
}
`;

exports[`GTF utils test parsing/formatting items and attributes 1`] = `
Object {
  "gene_id": Array [
    "\\"EDEN\\"",
  ],
  "gene_name": Array [
    "\\"EDEN\\"",
  ],
  "transcript_id": Array [
    "\\"EDEN.3\\"",
  ],
}
`;

exports[`GTF utils test parsing/formatting items and attributes 2`] = `"transcript_id EDEN.3; gene_id EDEN; gene_name EDEN;"`;
